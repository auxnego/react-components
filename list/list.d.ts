import { ReactNode, LIHTMLAttributes } from "react";

import {
  ClassNameConfig,
  Config
} from "../global.d";

/** Menu Component **/

export type MenuItemMainProps = {
 id:string|number,
 content:ReactNode,
 textContent?:string
};

export type MenuItemProps = LIHTMLAttributes<HTMLLIElement> &
                            MenuItemMainProps;

export type MenuMainProps = {
  items:MenuItemProps[],
  classPrefix?:string,
  onEmpty?:ReactNode,
  ordered?:boolean,
};

export type MenuProps = ClassNameConfig &
                        Config<
                          {root?:string, item?:string},
                          {root?:object, item?:object}, {}
                        > & 
                        MenuMainProps;
