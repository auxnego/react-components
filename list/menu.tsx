import React, { Attributes, ReactNode } from "react";
import clsx from "clsx";

//types
import {
  MenuItemProps,
  MenuProps
} from "./list.d";

export default function Menu(props:MenuProps){
  //default props
  let {
    classless = false,
    className = "",
    classPrefix = "menu",
    items,
    ordered = false,
    onEmpty = null,
    setClass = {root:"", item:""},
    setProps = {root:{}, item:{}},
    ...more
  } = props;

  //classNames
  let __classNames = {
    root:[!classless && {
      [`${classPrefix}__root`]:true
    }, className, setClass.root!],

    item:[!classless && {
      [`${classPrefix}__item`]:true
    }, setClass.item!]
  };

  //props
  let __props = {
    root:{
      ...more,
      ...setProps.root!,
      className:clsx(__classNames.root)
    } as Attributes,

    item:{
      ...setProps.item!,
      className:clsx(__classNames.item)
    } as Attributes
  };

  //views
  let __children:any;

  if(!items.length){
    if(onEmpty){
       __children = [<li {...__props.item} key="1">{onEmpty}</li>];
    }
  }else{
    __children = items.filter((e:any) => {
                     return !!Object.keys(e).length;
                   }).map((e:MenuItemProps, i:any) => {
                     let { content, textContent, ...more} = e;
                     return React.createElement("li", {...more, ...__props.item, key:i }, content);
                   }); 
  }

  //render
  return React.createElement(
    ordered ? "ol" : "ul",
    __props.root,
    __children
  );
}
