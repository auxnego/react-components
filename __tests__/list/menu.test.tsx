import React from "react";
import {create} from "react-test-renderer";

import { Menu } from "../../list";

let items = [{id:"i-1", content:"elemento 1"}, {id:2, content:React.createElement("i",{},"elemento2")}, {id:"id-3", content:"el3"}];

it("Menu without elements", () => {

  expect(create(
    <Menu items={[]} />
  ).toJSON()).toMatchSnapshot();

});

it("Menu without elements and empty indicator", () => {

  expect(create(
    <Menu items={[]} onEmpty="No hay Elementos" />
  ).toJSON()).toMatchSnapshot();

});

it("Menu with elements", () => {

  expect(create(
    <Menu items={items} />
  ).toJSON()).toMatchSnapshot();

});

it("Menu with classPrefix and ordered", () => {

  expect(create(
    <Menu items={items} classPrefix="custom" ordered/>
  ).toJSON()).toMatchSnapshot();

});
