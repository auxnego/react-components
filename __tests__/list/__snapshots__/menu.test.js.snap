// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu with classPrefix and ordered 1`] = `
<ol
  className="custom__root"
>
  <li
    className="custom__item"
    id="i-1"
  >
    elemento 1
  </li>
  <li
    className="custom__item"
    id={2}
  >
    <i>
      elemento2
    </i>
  </li>
  <li
    className="custom__item"
    id="id-3"
  >
    el3
  </li>
</ol>
`;

exports[`Menu with elements 1`] = `
<ul
  className="menu__root"
>
  <li
    className="menu__item"
    id="i-1"
  >
    elemento 1
  </li>
  <li
    className="menu__item"
    id={2}
  >
    <i>
      elemento2
    </i>
  </li>
  <li
    className="menu__item"
    id="id-3"
  >
    el3
  </li>
</ul>
`;

exports[`Menu without elements 1`] = `
<ul
  className="menu__root"
/>
`;

exports[`Menu without elements and empty indicator 1`] = `
<ul
  className="menu__root"
>
  <li
    className="menu__item"
  >
    No hay Elementos
  </li>
</ul>
`;
